app/src/main/java/com/example/sharingang/MapFragment.kt

```diff
@@ -62,41 +68,42 @@ class MapFragment : Fragment(), OnMapReadyCallback {
         binding = DataBindingUtil.inflate(inflater, R.layout.fragment_map, container, false)
         fusedLocationClient = LocationServices.getFusedLocationProviderClient(requireContext())
         hasCameraMovedOnce = false
+
         viewModel.searchResults.observe(viewLifecycleOwner, {
             addItemMarkers(it)
         })
+
         binding.mapGetMyLocation.setOnClickListener {
             if (lastLocation != null) {
                 moveCameraToLastLocation()
             }
         }
+
         binding.mapStartSearch.setOnClickListener {
             this.findNavController()
                 .navigate(MapFragmentDirections.actionMapFragmentToSearchFragment())
         }
+
         binding.mapView.onCreate(savedInstanceState)
-        binding.mapView.getMapAsync(this)
+
+        initMap()
+
         return binding.root
     }
 
     private fun addItemMarkers(items: List<Item>) {
-        map?.clear()
-        if (lastLocation != null) {
-            addLastLocationMarker()
-        }
-        for (item: Item in items) {
-            if (!item.sold) {
-                val addedMarker = map?.addMarker(
-                    MarkerOptions().position(LatLng(item.latitude, item.longitude))
-                        .icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_AZURE))
-                )
-                addedMarker?.tag = item
-            }
-        }
+        clusterManager?.clearItems()
+        clusterManager?.addItems(items.map { MapItem(it) })
+        clusterManager?.cluster()
     }
 
     @SuppressLint("MissingPermission")
     private fun startLocationUpdates() {
+        // Show blue dot at current location
+        map?.isMyLocationEnabled = true
+        // We already have a custom button so we remove the default one
+        map?.uiSettings?.isMyLocationButtonEnabled = false
+
```

Comment:
What happens if the location permission is not granted?

Reply:
I think that this function is only called when the permission has been granted.
I'm not sure, could you check that please? This function already existed before, I only added these 2 lines

Reply:
Oh okk my bad, then all good!